src/execution_algorithms/sgd_training_algorithm.cpp

```diff
@@ -83,6 +84,13 @@ void sgd_training_algorithm::train(sgd_execution_context& c,
                                    data_coordinator& dc,
                                    sgd_termination_criteria const& term)
 {
+  // Setup a "training-global" validation context:
+  using ValidationContext = sgd_execution_context;
+  static ValidationContext evaluation_context(
```

Comment:
Won't making this `static` prevent us from juggling multiple instances of `sgd_training_algorithm`?

Reply:
Why yes it will! But we don't have those. What we do have is legacy code in callbacks. It's easier to change these two lines than to try to fix callbacks. 🤷‍♂️